### ISSUE TITLE
VSCE - Ability to copy link to main branch rather than current

### ISSUE THREAD

**Issue by ryankscott** [sourcegraph/sourcegraph#34591]
> quickly pull up a shareable link to several lines in a file, or the whole file (“Open File in Sourcegraph Web” or “Copy Sourcegraph File Link”) ....
However, 99% of the time I’m doing this, I’m not on main, and I’d rather share the link on main.

It'd be great to be able to somehow have the option to copy a link to the main branch rather than the currently checked out one in VS Code.

**Comment by sourcegraph-bot-2** [comment]
Heads up @muratsu @jjinnii @ryankscott - the "team/integrations" label was applied to this issue.

**Comment by abeatrix** [comment]
@ryankscott we currently have a setting that allows users to set the default branch in their User setting, so files will always open at the default branch. 
However, I'm thinking we can improve the current workflow by adding a separate action to open file in `main` vs open in `current branch`.
We can also try updating the logic so that it would always open the file in `main` if we can detect the current branch doesn't exist remotely...?

Would like to hear your thoughts on this because it's also an issue I run into a lot 😄

**Comment by ryankscott** [comment]
My initial thought would be to have an option in dropdown so it's like:

```
    ┌────────────────────────┬────────────────────────┐
    │  Open in Sourcegraph > │ In current branch      │
    └────────────────────────┼────────────────────────┤
                             │ In main branch         │
                             └────────────────────────┘
```
However I'm not sure what the best UI convention is here.

More interestingly though, how do we *know* whether this will exist in the main branch? And if it does, how do we ensure the line numbers are correct etc.?

**Comment by jjinnii** [comment]
Quickly mocked up Ryan's idea + a second option in [Figma](https://www.figma.com/file/vEMPAnGxEO2QSn6STlWIbi/VSCE-Link-out-to-main-branch?node-id=275%3A1628). I think the deciding factor here is what's the expected behavior? 

If a developer clicks "open in sourcegraph", is it more often than not that they've forgotten that they haven't published the branch? Or if they do know that they haven't published, would they expect to open to main? 
cc: @abeatrix

**Comment by abeatrix** [comment]
> More interestingly though, how do we know whether this will exist in the main branch? 

I haven't really looked into it yet but maybe we can:
1. make an [API call](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/schema.graphql?L2576-2595) to get a list of the branches of that repo to check if it exists
2. or update our web app so that it would redirect us to the main branch directly instead of a 404 page if the branch doesn't exist? 
3. or use git command to check if it exists before we generate the link, but it's a bit slow to get the response back using this (if branch exists in remote we will get a response vs no reponse if it doesn't exist)
![image](https://user-images.githubusercontent.com/68532117/165858144-346cebbb-9c63-4e07-8c79-f34631935000.png)


> And if it does, how do we ensure the line numbers are correct etc.?

This is a good question, I don't think we can atm

> is it more often than not that they've forgotten that they haven't published the branch?

I think so! Speaking for myself but when I start working on a new feature or issue, I would start a new branch, and then that would be a branch that doesn't exist remotely yet but I wouldn't remember until I start opening files on Web :P 

> Or if they do know that they haven't published, would they expect to open to main?

For me I think any branch but 404, preferably main but I know there are customers that use different branches as their default branch for their team --that was the reason we added the defaultBranch setting 😆 

I hope this helps!

**Comment by jjinnii** [comment]
> This is a good question, I don't think we can atm

Hmm that's kind of a bummer then, so if a branch is published and a user navigates to Sourcegraph, we land them at the correct file path, but no specific line highlighted? 

>  I would start a new branch, and then that would be a branch that doesn't exist remotely yet but I wouldn't remember until I start opening files on Web 

The first option (Ryan's proposal) keeps us from re-directing the user _before_ they realize that they haven't published, which is good. But I wonder if the disabled (or missing) "go to current branch" button is clear that it's unavailable because they haven't published. I'm not keen on adding a help text for a sub menu in a menu popover 😆 

When I take that into consideration, the alert does a much better job of communicating _why_ a user is taken to main.

**Comment by ryankscott** [comment]
Sounds like that linking to main is going to be pretty unreliable to ensure we get users to the right place. Perhaps we should just ensure that someone can't share a link that isn't available in Sourcegraph.

Instead of disabling or removing the "Go to current branch" option, could we throw an alert when you click that and the branch is not published and not copy it to clipboard? We already use one when you copy a link from the command palette.

<img width="457" alt="Screen Shot 2022-04-29 at 13 17 33" src="https://user-images.githubusercontent.com/2660760/165934640-2dd807c7-cece-4865-a932-966b22fdb86b.png">

This would also handle the case where you use the command palette too.

**Comment by jjinnii** [comment]
Mocked up a [third option](https://www.figma.com/file/vEMPAnGxEO2QSn6STlWIbi/VSCE-Link-out-to-main-branch?node-id=1625%3A3800) based on Ryan's feedback here. This is specifically for the "Open File in Sourcegraph Web" flow, not the "Copy Sourcegraph File Link". 

Do we think we need to differentiate main/current branch for both the "go to sourcegraph" and "copy link" actions? If so, I have it mocked up in [here](https://www.figma.com/file/vEMPAnGxEO2QSn6STlWIbi/VSCE-Link-out-to-main-branch?node-id=1625%3A4673) for reference. I'm just using the same error message. 😄 

cc: @ryankscott @abeatrix

**Comment by abeatrix** [comment]
Looks good to me! 

@jjinnii @ryankscott question! what happens if I've set the defaultBranch in my setting? Should the `Open In main branch` button take me to the defaultBranch, or a separate option should show up if a default branch is configured?

**Comment by ryankscott** [comment]
In option 3 we remove the `Open In main branch` button, consequently the logic should stay the same. For clarity that is:
 - If a default branch is **set**, use that
 - If a default branch is **not set** use the current branch

We will still need the alerts there, in case either your current or default branch isn't shared.

I just realised that in any case, if you haven't pushed your changes we can't reliably recreate a link anyway. Did we just ignore this case for the first iteration? Perhaps in future we can have an alert notifying if people have not pushed any local changes also.

**Comment by abeatrix** [comment]
Final Implementation will be included in the next release. https://github.com/sourcegraph/sourcegraph/pull/35519/

Behavior as described by Sara:
```
All secondary menus for "Go/link to current branch" and "Go/link to main branch" have been removed, and there are two behaviors:

If current branch exists, user is taken directly to sourcegraph.com
If current branch does not exist, action triggers notification
```

Loom: https://www.loom.com/share/edd9c28e828640a1b803e6197af6dfdc

**Comment by daprahamian** [comment]
I feel like this broke my current workflow. I NEVER want to open on the current branch I am on, I always want to open on the default branch. Is it possible to achieve that now?

**Comment by abeatrix** [comment]
@daprahamian Yes! If a default branch is set in your user setting using the `sourcegraph.defaultBranch` option, eg `"sourcegraph.defaultBranch": "main"`, it will always open the file on defaultBranch. 

Please make sure you are on v2.2.4 for the extension to behavior correctly, thanks!

**Comment by daprahamian** [comment]
@abeatrix Ah, yes, perfect! Yup, appears that i was on 2.2.3, upgrading to 2.2.4 fixed it. Thank you!
